src/entity.ts

```diff
@@ -383,49 +384,101 @@ export class Entity {
 	}
 }
 
+export type EntityPropertiesOfType<T> = {
+    [P in keyof T]:
+		T[P] extends (infer TItem)[]
+			? ObservableArray<(
+				TItem extends string ? string :
+				TItem extends number ? number :
+				TItem extends boolean ? boolean :
+				TItem extends Date ? Date :
+				EntityOfType<TItem>
+			)>
+			: (
+				T[P] extends string ? string :
+				T[P] extends number ? number :
+				T[P] extends boolean ? boolean :
+				T[P] extends Date ? Date :
+				EntityOfType<T[P]>
+			) | null;
+};
+
+type EntityOfTypeMemberOverrides<T> = {
+    update(args: EntityArgsOfType<T>): Promise<void>;
+};
+
+export type EntityOfType<T> = EntityPropertiesOfType<T> & EntityOfTypeMemberOverrides<T> & Entity; // Omit<Entity, keyof EntityOfTypeMemberOverrides<T>> & ;
+
+export type TEntityConstructor<T> = {
+    new(id: string, args?: EntityArgsOfType<T>): EntityOfType<T>;
+    new(args?: EntityArgsOfType<T>): EntityOfType<T>;
+	meta: TypeOfType<T>;
+};
+
+export type EntityArgsOfType<T> = Partial<{
+    -readonly [P in keyof T]:
+		T[P] extends (infer TItem)[]
+			? (
+				TItem extends string ? string[] :
+				TItem extends number ? number[] :
+				TItem extends boolean ? boolean[] :
+				TItem extends Date ? Date[] :
+				(EntityOfType<TItem> | Partial<EntityArgsOfType<TItem>>)[]
+			)
+			: (
+				T[P] extends string ? string :
+				T[P] extends number ? number :
+				T[P] extends boolean ? boolean :
+				T[P] extends Date ? Date :
+				(EntityOfType<T[P]> | Partial<EntityArgsOfType<T[P]>>)
+			) | null;
+}>;
+
+// TODO: Replace with TEntityConstructor
 export interface EntityConstructor {
 	new(): Entity;
 	new(properties?: ObjectLookup<any>): Entity; // Construct new instance with state
 }
 
-export interface EntityConstructorForType<TEntity extends Entity> extends EntityConstructor {
-	new(): TEntity;
-	new(properties?: ObjectLookup<any>): TEntity; // Construct new instance with state
+// TODO: Replace with TEntityConstructor, rename to EntityConstructorOfType
+export interface EntityConstructorForType<EntityType extends Entity> extends EntityConstructor {
+	new(): EntityType;
+	new(properties?: ObjectLookup<any>): EntityType; // Construct new instance with state
 	meta: Type;
 }
 
-export interface EntityRegisteredEventArgs {
-	entity: Entity;
+export interface EntityRegisteredEventArgs<EntityType extends Entity> {
```

Comment:
_info_: Add generic type argument `EntityType` to a lot of types so that they can leverage more specific typing when needed.